File: minesweeper/java-minesweeper/src/test/java/it/chicio/minesweeper/field/parser/FieldsParserByDivingInputIntoRowsTest.java
package it.chicio.minesweeper.field.parser;

import it.chicio.minesweeper.field.Field;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentMatchers;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static java.util.Arrays.asList;
import static org.hamcrest.CoreMatchers.*;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static org.mockito.hamcrest.MockitoHamcrest.argThat;

public class FieldsParserByDivingInputIntoRowsTest {

    private FieldRowParser fieldRowParser;
    private FieldsParserByDivingInputIntoRows fieldsParserByDivingInputIntoRows;

    @Before
    public void setUp() {
        fieldRowParser = mock(FieldRowParser.class);
        fieldsParserByDivingInputIntoRows = new FieldsParserByDivingInputIntoRows(fieldRowParser);
        when(fieldRowParser.parse(argThat(not("0 0")), ArgumentMatchers.any(FieldsParsingStatus.class)))
                .thenReturn(new FieldsParsingStatusBuilder().build());
    }

    @Test
    public void parseTermination() {
        when(fieldRowParser.parse(argThat(is("0 0")), ArgumentMatchers.any(FieldsParsingStatus.class)))
                .thenReturn(new FieldsParsingStatusBuilder().withFieldsParsed(new ArrayList<Field>()).build());
        FieldsParserByDivingInputIntoRows fieldsParserByDivingInputIntoRows =
                new FieldsParserByDivingInputIntoRows(fieldRowParser);

        List<Field> fields = fieldsParserByDivingInputIntoRows.parse("0 0");

        assertThat(fields, is(equalTo(Collections.<Field>emptyList())));
    }

    @Test
    public void parseAValidField() {
        when(fieldRowParser.parse(argThat(is("0 0")), ArgumentMatchers.any(FieldsParsingStatus.class)))
                .thenReturn(new FieldsParsingStatusBuilder().withFieldsParsed(
                                new ArrayList<Field>(Collections.singletonList(new Field(new String[][]{{"*"}})))
                ).build()
        );

        List<Field> fields = fieldsParserByDivingInputIntoRows.parse(
                "1 1" + System.getProperty("line.separator") +
                "*" + System.getProperty("line.separator") +
                "0 0"
        );

        assertThat(fields, is(equalTo(asList(new Field(new String[][]{{"*"}})))));
    }

    @Test
    public void parseTwoValidFields() {
        when(fieldRowParser.parse(argThat(is("0 0")), ArgumentMatchers.any(FieldsParsingStatus.class)))
                .thenReturn(
                new FieldsParsingStatusBuilder().withFieldsParsed(
                        new ArrayList<Field>(asList(new Field(new String[][]{{"*"}}), new Field(new String[][]{{"*"}})))
                ).build()
        );

        List<Field> fields = fieldsParserByDivingInputIntoRows.parse(
                "1 1" + System.getProperty("line.separator") +
                "*" + System.getProperty("line.separator") +
                System.getProperty("line.separator") +
                "1 1" + System.getProperty("line.separator") +
                "*" + System.getProperty("line.separator") +
                "0 0"
        );

        assertThat(fields, is(equalTo(asList(
                new Field(new String[][]{{"*"}}),
                new Field(new String[][]{{"*"}})
        ))));
    }
}

File: minesweeper/kotlin-mockk-minesweeper/src/test/kotlin/it/chicio/minesweeper/field/parser/FieldsParserByDivingInputIntoRowsTest.kt
package it.chicio.minesweeper.field.parser

import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.verify
import it.chicio.minesweeper.FieldFactory
import it.chicio.minesweeper.FieldsParsingStatusBuilder
import it.chicio.minesweeper.field.Field
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import java.util.*

@DisplayName("FieldsParserByDivingInputIntoRows")
@ExtendWith(MockKExtension::class)
class FieldsParserByDivingInputIntoRowsTest {
    @Nested
    @DisplayName("parse")
    inner class Parse {
        @MockK
        private lateinit var fieldRowParser: FieldRowParser

        private lateinit var fieldsParserByDivingInputIntoRows: FieldsParserByDivingInputIntoRows

        @BeforeEach
        fun setUp() {
            fieldsParserByDivingInputIntoRows = FieldsParserByDivingInputIntoRows(fieldRowParser)
        }

        @Test
        fun termination() {
            every { fieldRowParser.parse(eq("0 0"), any()) } returns FieldsParsingStatusBuilder().build()

            val fields = fieldsParserByDivingInputIntoRows.parse("0 0")

            assertEquals(fields, emptyList<Field>())
            verify(exactly = 1) { fieldRowParser.parse(eq("0 0"), any()) }
        }

        @Test
        fun `a valid field`() {
            every { fieldRowParser.parse(eq("0 0"), any()) } returns FieldsParsingStatusBuilder(
                    fieldsParsed = ArrayList(listOf(FieldFactory().make(arrayOf(arrayOf("*")))))
            ).build()
            every { fieldRowParser.parse(neq("0 0"), any()) } returns FieldsParsingStatusBuilder().build()

            val fields = fieldsParserByDivingInputIntoRows.parse(
                    """
                1 1
                *
                0 0
                """.trimIndent()
            )

            assertEquals(fields, listOf(FieldFactory().make(arrayOf(arrayOf("*")))))
            verify(exactly = 1) { fieldRowParser.parse(eq("0 0"), any())  }
            verify { fieldRowParser.parse(neq("0 0"), any())  }
        }

        @Test
        fun `two valid fields`() {
            every { fieldRowParser.parse(eq("0 0"), any()) } returns FieldsParsingStatusBuilder(
                    fieldsParsed = ArrayList(listOf(FieldFactory().make(arrayOf(arrayOf("*"))), FieldFactory().make(arrayOf(arrayOf("*")))))
            ).build()
            every { fieldRowParser.parse(neq("0 0"), any()) } returns FieldsParsingStatusBuilder().build()

            val fields = fieldsParserByDivingInputIntoRows.parse(
                    """
                1 1
                *
                
                1 1
                *
                0 0
                """.trimIndent()
            )
            assertEquals(fields, listOf(
                    FieldFactory().make(arrayOf(arrayOf("*"))),
                    FieldFactory().make(arrayOf(arrayOf("*")))
            ))
            verify(exactly = 1) { fieldRowParser.parse(eq("0 0"), any())  }
            verify { fieldRowParser.parse(neq("0 0"), any())  }
        }
    }
}